Add CleanUpInterpolator.

markscheel opened the pull request:
This is a PR towards parallel interpolation.

This is independent of #1075 and #1082 (and should be easier to review than those).

### Types of changes:

- [x] New feature

### Component:

- [x] Code
### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel committed:
Add CleanUpInterpolator.

This is a PR towards parallel interpolation.

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/CleanUpInterpolator.hpp:
```diff
@@ -0,0 +1,119 @@
+        [&temporal_id](
+            const gsl::not_null<db::item_type<
+                Tags::InterpolatedVarsHolders<Metavariables, VolumeDim>>*>
+                holders) {
```
`noexcept`

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/CleanUpInterpolator.hpp:
```diff
@@ -0,0 +1,119 @@
+    const auto& holders =
+        db::get<Tags::InterpolatedVarsHolders<Metavariables, VolumeDim>>(box);
+    tmpl::for_each<typename Metavariables::interpolation_target_tags>(
+        [&](auto tag) {
```
`noexcept`

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/CleanUpInterpolator.hpp:
```diff
@@ -0,0 +1,119 @@
+          make_not_null(&box),
+          [&temporal_id](const gsl::not_null<db::item_type<
+                             Tags::VolumeVarsInfo<Metavariables, VolumeDim>>*>
+                             volume_vars_info) {
```
`noexcept`

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/CleanUpInterpolator.hpp:
```diff
@@ -0,0 +1,119 @@
+          [&temporal_id](
+              const gsl::not_null<db::item_type<
+                  Tags::InterpolatedVarsHolders<Metavariables, VolumeDim>>*>
+                  holders_l) {
```
`noexcept`

nilsdeppe commented on src/NumericalAlgorithms/Interpolation/CleanUpInterpolator.hpp:
```diff
@@ -0,0 +1,119 @@
+                  Tags::InterpolatedVarsHolders<Metavariables, VolumeDim>>*>
+                  holders_l) {
+            tmpl::for_each<typename Metavariables::interpolation_target_tags>(
+                [&](auto tag) {
```
`noexcept`

nilsdeppe reviewed (changes requested):
I'm happy for you to squash the minor comments I have :+1:

wthrowe commented on src/NumericalAlgorithms/Interpolation/CleanUpInterpolator.hpp:
```diff
@@ -0,0 +1,119 @@
+              } else {
+                ++it;
+              }
+            }
```
Isn't this loop just `volume_vars_info->erase(temporal_id);`?

markscheel committed:
IWYU changes in non-cpp files.

These are files from previous PRs that IWYU was never
run on.  Travis does not run IWYU on these files.

markscheel commented:
Fixed loop, and added a new commit where I fixed includes according to a manual IWYU run on files that Travis doesn't run IWYU on.

nilsdeppe reviewed (approved):
@wthrowe could you please look at this when you get a chance? Thanks :)

nilsdeppe commented:
Restart GCC job that failed with random segfaults.
